test/realm_test.dart

```diff
@@ -1567,6 +1566,75 @@ Future<void> main([List<String>? args]) async {
     //test the realm can be opened.
     final realm = getRealm(Configuration.disconnectedSync([Product.schema], path: path, encryptionKey: key));
   });
+
+  test('Realm.refresh', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final result = realm.refresh();
+    expect(result, true);
+  });
+
+  test('Realm.onRefresh sync transaction', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    var called = false;
+    final transaction = realm.beginWrite();
+    realm.refreshAsync().then((_) => called = true);
+    realm.add(Person("name"));
+    transaction.commit();
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+    expect(called, true, reason: "beginWrite onRefresh failed");
+
+    called = false;
+    realm.write(() {
+      realm.refreshAsync().then((_) => called = true);
+      realm.add(Person("name"));
+    });
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+    expect(called, true, reason: "write onRefresh failed");
+  });
+
+  test('Realm.onRefresh async transaction', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    bool called = false;
+    final transaction = await realm.beginWriteAsync();
+    realm.refreshAsync().then((_) => called = true);
+    realm.add(Person("name"));
+    await transaction.commitAsync();
+    expect(called, true);
+  });
+
+   test('Realm.onRefresh is called if registered outside a transaction', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    bool called = false;
+
+    realm.refreshAsync().then((_) => called = true);
+
+    realm.beginWrite().commit();
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+
+    expect(called, true);
+
+
+    realm.refreshAsync().then((_) => called = true);
+
+    realm.write(() {});
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+
+    expect(called, true);
+  });
+
+   test('Realm.onRefresh on frozen realm should be no-op', () async {
+    var realm = getRealm(Configuration.local([Person.schema]));
+    bool called = false;
+    realm = realm.freeze();
+
+    await Future.any([realm.refreshAsync().then((_) => called = true), Future<void>.delayed(Duration(milliseconds: 10))]);
```

Comment:
Probably you could use `waitFutureWithTimeout`.

Reply:
This is test specific. The problem with these functions is they become part of the internal testing API. And here problem is highlighted. I prefer to use the common Future API instead of increasing the size of the internal API which then one needs to remember that it exists